CMakeLists.txt

```
@@ -165,6 +165,18 @@ if(BUILD_CLANG_PLUGIN)
   if(NOT EXISTS ${CLANG_INCLUDE_PATH})
     message(SEND_ERROR "clang include path ${CLANG_INCLUDE_PATH} does not exist. Please provide clang's internal include path manually: Find the directory where __clang_cuda_runtime_wrapper.h is. Provide this directory for older ROCm versions and the parent directory for newer ones.")
   endif()
+  if(WITH_ROCM_BACKEND)
+    execute_process(COMMAND ${CLANG_EXECUTABLE_PATH} "--version"
+      OUTPUT_VARIABLE CLANG_VERBOSE_OUTPUT)
+    string(REGEX MATCH "clang version [0-9]+.[0-9]+.[0-9]+ \\(.+ roc-([0-9]+).([0-9]+).([0-9]+)" AMD_CLANG_VERSION ${CLANG_VERBOSE_OUTPUT})
+
+    if(AMD_CLANG_VERSION)
+      set(ROCM_VERSION_MAJOR ${CMAKE_MATCH_1})
+      set(ROCM_VERSION_MINOR ${CMAKE_MATCH_2})
+      set(ROCM_VERSION_PATCH ${CMAKE_MATCH_3})
+      message(STATUS "AMD clang version: ${ROCM_VERSION_MAJOR}.${ROCM_VERSION_MINOR}.${ROCM_VERSION_PATCH}")
+    endif()
+  endif()
```

Would also be nice to automatically enable the device mangler fix:

```suggestion
  endif()
  if(${LLVM_VERSION_MAJOR} EQUAL 13 AND AMD_CLANG_VERSION AND NOT DEFINED ${HIPSYCL_NO_DEVICE_MANGLER})
      message(STATUS "Enabling workarounds for AMD clang 13.x.x: HIPSYCL_NO_DEVICE_MANGLER")
      set(HIPSYCL_NO_DEVICE_MANGLER TRUE)
  endif()
```

---

It's not really a fix though, just a workaround at the expense of other functionality... I don't think this should be applied automatically because it *will* break any explicit multipass compilation flow, leading to silent kernel launch failures due to mismatch of the mangled kernel name between the host binary and the compiled device binary.

So, this should probably be opt-in, with clear warning signs...

---

If we have Clang 13 from ROCm 4.5.2 (and likely earlier versions as well), hipSYCL would not compile unless the device mangler functionality is disabled. As long as we trust the version detection, I don't see how **not** setting `HIPSYCL_NO_DEVICE_MANGLER` is better for functionality.

But if you think emitting a warning is better, that's fine too. Certainly better than the compile error :)

Alternatively, the most robust approach, IMHO, would be to use `check_cxx_source_compiles`, but it requires carefully setting the flags to use the correct LLVM. I tried a bit and have not managed to make it work.

---

I just added an error in CMake (using your condition) if `HIPSYCL_NO_DEVICE_MANGLER' is not set, as it'd error in the compilation anyways.

---

@al42and The issue is that it will break half of hipSYCL's compilation flows in a very subtle and hard-to-debug way. And cmake warnings are often not something that sticks out very much.

We've decided that the best course of action is to inform the user that  they need to set `HIPSYCL_NO_DEVICE_MANGLER` in case we detect the problem. In that way, the user forced to explicitly acknowledge that they do not want explicit multipass support, including all of the consequences.

---

Ok, makes sense.